CONTRIBUTING.md

```diff
@@ -161,6 +161,42 @@ the docs will be built under `doc/_build/html`. Detail information about documen
 ## CI
 All PRs triger a CI job to run linting, type checking, tests, and build docs. The CI script is located [here](https://github.com/SeldonIO/alibi/blob/master/.github/workflows/ci.yml) and should be considered the source of truth for running the various development commands.
 
+## Optional Dependencies
+
+Alibi uses optional dependencies to allow users the choice to not install large or error-prone dependencies. These are 
+managed in the `requirements/extra.txt` file as in inverse index. Alibi manages modularity of components that depend on 
+optional dependencies using the `import_optional` defined in `alibi/utils/missing_optional_dependency.py`. This 
+replaces the dependency with a dummy class that raises an error when called. If you are working on public functionality 
+that is dependent on an optional dependency you should expose the functionality via the relevant `__init__.py` file by
+importing it there using the `optional_import` function. Currently, optional dependencies are tested by importing all 
+the public functionality and checking that the correct errors are raised dependent on the environment. Developers can 
+run these tests using `tox`. These tests are in `alibi/tests/test_dep_mangement.py`. If implementing functionality that 
+is dependent on a new optional dependency then you will need to:
+
+1. Add it to `requirements/extra.txt`.
+2. Create a new `tox` environment in `setup.cfg` with the new dependency.
+3. Define a new MissingDependency class in `alibi/utils/missing_optional_dependency.py` and integrate it with the 
+   `import_optional` function.
+4. Make sure any public functionality is protected by the `import_optional` function.
+5. Make sure the new dependency is tested in `alibi/tests/test_dep_mangement.py`.
+
+Note that subcomponents can be dependent on optional dependencies too. In this case the user should be able to import 
+and use the relevant parent component. The user should only get an error message if: 
+1. They don't have the optional dependency installed and,
+2. They configure the parent component in such as way that it uses the subcomponent functionality.
+
+Developers should implement this by importing the subcomponent into the source code defining the parent component using 
+the `optional_import` function. To see an example of this look at the `AnchorText` and `LanguageModelSampler` 
+subcomponent implementation.
+
+#### Note:
+- The `import_optional` function mirrors the python import functionality and thus its return type has to be `Any`. Using 
+objects imported with this function can lead to misspecification of types as `Any` when the developer intended to be 
+more restrictive. If you want to type a variable using a class that depends on an optional dependency then you should 
+use the `TYPE_CHECKING` to import it instead.
```

Comment:
Perhaps need to make it clearer, i.e. using the `TYPE_CHECKING` guard directly import the class (not using `import_optional`) as all optional deps would be present in a dev environment.

Reply:
It's also important to spell out that in the case of `TYPE_CHECKING` the imports have to be from the underlying modules which implement the classes rather than the public API due to the `Any` return type of the latter with an example.